### PR TITLE
handle remove from sale

### LINKED ISSUE #1754: Cancel NFT sale from NFT tile
**AC**

- [ ] [Remove from sale](https://www.figma.com/file/yhZpTHdf1sxJx13uRZ71GV/Membership-profile?node-id=972%3A224906)

### PR DESCRIPTION
**Pull request by rafalpawlow**
Closes #1754 

**Design**
https://www.figma.com/file/yhZpTHdf1sxJx13uRZ71GV/Membership-profile?node-id=972%3A218668

### CODE REVIEW

File: packages/atlas/src/components/NftTile/NftTileDetails.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTile/NftTileDetails.tsx]: I think those shouldn't be in this component. So far both `NftTile` and `NftTileDetails` have been dumb components, only working based on props received. This makes them smart and quite a lot. We probably should pass all the necessary data via props and include all the logic/hooks in `NftTileViewer` instead

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: I think this could accept the nft data directly, the user of this hook should be responsible for fetching it already

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: What's the point of all the `useMemo` here? We don't do any complicated operations and we have stable references anyway since all of those are booleans

**Review by kdembler — CHANGES_REQUESTED** [review]

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts:20]: Probably needs to include the member as well

File: packages/atlas/src/api/hooks/nfts.ts
Comment (drillprop) [inline packages/atlas/src/api/hooks/nfts.ts:16]: Can we also add here type of auction - i.e. `open-auction` and `english-auction`?

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (drillprop) [inline packages/atlas/src/components/NftTile/NftTile.tsx]: I'm aware that it's not your change, but I'm confused here. What means `on-sale`? Is it `buy-now`? If yes, can we make this consistent with changes in `useNft`?

File: packages/atlas/src/api/hooks/nfts.ts
Comment (drillprop) [inline packages/atlas/src/api/hooks/nfts.ts]: typo:
```suggestion
    nftStatus: getNftProperties(),
```

File: packages/atlas/src/api/hooks/nfts.ts
Comment (drillprop) [inline packages/atlas/src/api/hooks/nfts.ts]: typo:
```suggestion
  const getNftProperties = (): NftStatus => {
```

**Review by drillprop — CHANGES_REQUESTED** [review]

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (eldiegod) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: yes `on-sale` means that you can buy it now

Comment (drillprop) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: hmm, so I think `buy-now` will be a much better name.

File: packages/atlas/src/api/hooks/nfts.ts
Comment (eldiegod) [inline packages/atlas/src/api/hooks/nfts.ts]: Why do we need auctionPlannedEndBlock on a non auction state?

File: packages/atlas/src/api/hooks/nfts.ts
Comment (eldiegod) [inline packages/atlas/src/api/hooks/nfts.ts]: same as above

File: packages/atlas/src/components/_nft/NftWidget/NftWidget.tsx
Comment (eldiegod) [inline packages/atlas/src/components/_nft/NftWidget/NftWidget.tsx]: If these last 3 props are common to all nft statuses we can just move them up to the general NftWidgetProps

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: This will be set only for finished auction which should never make their way into active transactional status, I think we can remove this field

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts:24]: Those aren't included in other variants of this type

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: Same here, not used

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (eldiegod) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: so I don't agree because the state in which the nft is is called being on sale the buy now is the action you can take, is like if you call the auction `bid-now`, but granted that we have `TransactionalStatusBuyNow` then i guess it would make sense to call it the same as that 🤷‍♂️

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: Planned end doesn't exist for `idle` state

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: Planned end doesn't exist for `buy-now` state

File: packages/atlas/src/components/_nft/NftWidget/NftWidget.tsx
Comment (kdembler) [inline packages/atlas/src/components/_nft/NftWidget/NftWidget.tsx:284]: Seems like almost everything that is in this hook should now just rely on `useNftState`?

**Review by eldiegod — CHANGES_REQUESTED** [review]
There is a couple more states from the nfts that we are going to need, I suggest taking a look at my PR for the updated types
https://github.com/Joystream/atlas/pull/2304/files#diff-b713a61de81d4417e5d506e571b7f2a11d0f02ce71f84a797faf398b22e9bf52

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (drillprop) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: I'm not sure if I understand. So, you're saying that `on-sale` means either `TransactionalStatusBuyNow ` or `TransactionalStatusAuction` **but with buy now price set**, right? 
If yes, this is very confusing. Maybe we could differentiate this even more - status could be `idle` or `only-buy-now` or `auction-buy-now` or `auction`?

**Comment by eldiegod** [comment]
no, I mean that the states that the Nft can be in are `idle` you can't do
much, `on-sale` you can buy now for a fixed price; and `auction` you can
bid and also buy now depending on params.

What I mean is that the wording `buy-now` is not a state of the nft itself
but an action taken by the user just like bidding is and `auction` would be
the state, does that make sense?

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (kdembler) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: I think we should reflect transactional status. So `idle`, `auction`, `buy-now`, I think it's quite clear. Potentially we could split auction into separate auction types if we wanted but up to discussion

**Comment by kdembler** [comment]
Diego, `buy-now` is a state though, as defined by runtime and query node. I agree it may not be the most fortunate name, but I think it may be the best to follow that

**Comment by eldiegod** [comment]
> s defined by runtime and query node. I agree it may not be the most fortunate name, but I think it may be the best to follow that

Yeah, that's what I meant 2 comments back, it's called like that in runtime and query node but it's confusing as well as you can also `buy-now` in auction. I don't mind just making it clear

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (eldiegod) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: I wouldn't split auction but add a type inside of 'open' | 'english', did this in my PR with the new changes

Comment (kdembler) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: That works as well

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: `duration` and `title` are missing from those variants but are returned in the data. If they are needed, let's include them

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTile/NftTile.tsx]: We use `on*` for props (we should at least 😅)

File: packages/atlas/src/components/NftTile/NftTileDetails.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTile/NftTileDetails.tsx]: Same here, `onRemoveFromSale`

File: packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx]: Would make most sense for this modal to live in some kind of hook so that it can be reused for widget/tile/etc.

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: English auction can be cancelled only if no bids have been made yet

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: You define those handy vars but later in the code keep checking the full `nft?.transactionalStatus?.__typename`, let's make use of them

File: packages/atlas/src/hooks/useNftTransactions.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftTransactions.ts]: Could this go into the `useNftActions` hook as well?

Comment (kdembler) [inline packages/atlas/src/hooks/useNftTransactions.ts]: Should use `useCallback`

**Review by kdembler — CHANGES_REQUESTED** [review]
Shaping up, really like it

File: packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx
Comment (kdembler) [reply packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx]: It can be a part of `cancelNftSale`

File: packages/atlas/src/components/NftTile/NftTile.tsx
Comment (rafalpawlow) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: Sorry, it's not the first time. I was in the project before where we had rule for `handle*` 😅

Comment (kdembler) [reply packages/atlas/src/components/NftTile/NftTile.tsx]: No worries. `on*` for props, `handle*` for handlers

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: Do we need this type? I'm afraid it may be a bit ambigous what it represents

File: packages/atlas/src/api/hooks/nfts.ts
Comment (kdembler) [inline packages/atlas/src/api/hooks/nfts.ts]: If `buyNowPrice` is missing, this will produce `NaN`. You could fix it by changing to `||` instead of `??`

File: packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx]: nft ID and video ID are the same, you need only single one

File: packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx
Comment (kdembler) [inline packages/atlas/src/components/NftTileViewer/NftTileViewer.tsx]: `handleRemoveFromSale` :D

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: This should also include buy now status

File: packages/atlas/src/hooks/useNftTransactions.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftTransactions.ts]: This isn't needed

File: packages/atlas/src/hooks/useNftState.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftState.ts]: Only owner can cancel sale

File: packages/atlas/src/hooks/useNftTransactions.ts
Comment (kdembler) [inline packages/atlas/src/hooks/useNftTransactions.ts:19]: Let's add `return` here. If you don't return the promise, then it cannot be awaited and `useTransaction` won't be able to catch error if it happens

**Review by kdembler — APPROVED** [review]
Leaving an approve here so my review doesn't block merging. However, let's get approval from @drillprop and @eldiegod before merging

**Review by drillprop — APPROVED** [review]
Looks good to me! Great work

**Review by eldiegod — APPROVED** [review]
Looks good, maybe we can revisit the part about having 3 hooks for nfts later

**Pull request merged**
